### ISSUE TITLE
Improve performance of consumer-group routes

### ISSUE THREAD

**Issue by msaggar** [cloudhut/kowl#292]
I have a large cluster of ~45k consumer groups. Opening the `Consumer Group` tab in the frontend times out in the 25 second default that `fetch` waits for responses from the backend. I can see in debug logs that the operation takes about 5 minutes to complete the request to `GET /api/consumer-groups` for my cluster.

**Comment by msaggar** [comment]
Moving discussion with @weeco from discord to this thread:

From @weeco  - 
```The only place where we use websockets would be the messages explore feature.
We have several options to handle this I guess. One option could be that we submit a second request from the frontend to describe the consumer groups we see on each page

So:
1. List all consumer groups => Send consumer group names to the frontend
2. Frontend describes all consumer groups that are visible on the currently selected page

Usually in webdev one would use cursors for pagination but Kafka doesn't have a cursor for us here```

Right now we are pulling about 2MB of data from just grabbing consumer groupIDs in our test cluster. As we get to higher environments, the data will get bigger but I am not sure by how much. If we go the route recommended of pulling all cgIDs, storing them locally on the frontend I would imagine browsers might not like that. @weeco any worries with that approach?

**Comment by weeco** [comment]
@msaggar Thanks for the details, 2-4MB is indeed really a lot, but I think in regards to the browser performance not yet a problem (we handle larger data in the explore messages tab I think). However you are right that it doesn't scale infinitely. May I ask why you have 50k+ consumer groups? (Just curious how one can end up with that many consumer groups tbh).

If just the consumer group names itself are already too big, there's no way around doing pagination that is supported by the backend. Our current pagination is fake in the sense that the frontend receives all data and just renders up to 200 items per page. I'd need to think about how we could implement pagination as this would make our Kowl instances somehow stateful - Kafka itself doesn't support the concept of pagination like databases do. Saying that - introducing a database as result cache that would allow us to do pagination is one option 😆 .

**Comment by msaggar** [comment]
The 50k+ consumer groups is likely us not using kafkajs correctly. We start a new consumer for each pod in our service. As pods come in and out we get to this number.

And as far as storing some state with Kowl by introducing a DB or cache, I see there is a ticket here: https://github.com/cloudhut/kowl/issues/20

**Comment by weeco** [comment]
I feel like only showing the consumer group names in that page will be very disappointing. How about this:

1. Return all consumer group names to the frontend
2. The frontend can request all consumer group details in bulk for each page (may require a new backend endpoint).

Do you want to proceed with coding this (you could also focus on just backend or frontend if you are just comofrtable with one part of the stack)?

**Comment by weeco** [comment]
@msaggar I had a look today and figured out (together with Travis) that the Kafka client stacks lots of requests without opening further connections to send the many admin requests concurrently. Thus it takes longer than necessary until all consumer group offsets are fetched. The Kafka client (franz-go) will be patched, so that it is controllable how many requests can be sent concurrently per broker, so that more requests can be handled concurrently.

That would probably work for 99% of all users, but for users with more than 5k Consumer Groups the response size becomes problematic. Therefore we still have to involve the frontend to address this.

**Comment by msaggar** [comment]
@weeco yeah i agree, we are taking investigating the kafka library we are using to see if there is anyway we can reduce the number of consumer groups. 

Do we want to continue with the change discussed above?

**Comment by msaggar** [comment]
@weeco we lowered our offset retention and now have ~2k consumers so I think this should be fine
